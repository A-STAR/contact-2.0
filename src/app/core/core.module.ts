import { ErrorHandler, NgModule, Optional, SkipSelf } from '@angular/core';
import { DatePipe } from '@angular/common';

import { JwtHelper } from 'angular2-jwt';
import { TranslateService } from '@ngx-translate/core';
import { StoreModule } from '@ngrx/store';
import { EffectsModule } from '@ngrx/effects';
import { StoreDevtoolsModule } from '@ngrx/store-devtools';

import { environment } from '../../environments/environment';

import { AuthEffects } from './auth/auth.effects';
import { AuthHttpService } from './auth/auth-http.service';
import { AuthService } from './auth/auth.service';
import { ContentTabService } from '../shared/components/content-tabstrip/tab/content-tab.service';
import { DataService } from './data/data.service';
import { DictionariesEffects } from './dictionaries/dictionaries.effects';
import { DictionariesService } from './dictionaries/dictionaries.service';
import { EntityTranslationsService } from './entity/translations/entity-translations.service';
import { ErrorHandlerService } from './error/error-handler.service';
import { LookupEffects } from './lookup/lookup.effects';
import { LookupService } from './lookup/lookup.service';
import { MenuService } from './menu/menu.service';
import { MetadataService } from './metadata/metadata.service';
import { NotificationsEffects } from './notifications/notifications.effects';
import { NotificationsService } from './notifications/notifications.service';
import { SettingsService } from './settings/settings.service';
import { ThemesService } from './themes/themes.service';
import { throwIfAlreadyLoaded } from './module-import-guard';
import { MetadataEffects } from './metadata/metadata.effects';
import { UserConstantsEffects } from './user/constants/user-constants.effects';
import { UserConstantsService } from './user/constants/user-constants.service';
import { UserDictionariesEffects } from './user/dictionaries/user-dictionaries.effects';
import { UserDictionariesService } from './user/dictionaries/user-dictionaries.service';
import { UserLanguagesEffects } from './user/languages/user-languages.effects';
import { UserLanguagesService } from './user/languages/user-languages.service';
import { UserPermissionsEffects } from './user/permissions/user-permissions.effects';
import { UserPermissionsService } from './user/permissions/user-permissions.service';
import { ValueConverterService } from './converter/value/value-converter.service';

import { rootReducer } from './state/root.reducer';

@NgModule({
  imports: [
    StoreModule.provideStore(rootReducer),
    EffectsModule.run(AuthEffects),
    EffectsModule.run(DictionariesEffects),
    EffectsModule.run(LookupEffects),
    EffectsModule.run(NotificationsEffects),
    EffectsModule.run(UserConstantsEffects),
    EffectsModule.run(UserDictionariesEffects),
    EffectsModule.run(UserLanguagesEffects),
    EffectsModule.run(UserPermissionsEffects),
    EffectsModule.run(MetadataEffects),
    environment.production
      ? []
      : StoreDevtoolsModule.instrumentOnlyWithExtension({
          maxAge: 1024
        })
  ],
  providers: [
    AuthHttpService,
    AuthService,
    ContentTabService,
    EntityTranslationsService,
    DictionariesService,
    DatePipe,
    DataService,
    JwtHelper,
    LookupService,
    MenuService,
    MetadataService,
    NotificationsService,
    SettingsService,
    ThemesService,
    TranslateService,
    UserConstantsService,
    UserDictionariesService,
    UserLanguagesService,
    UserPermissionsService,
    ValueConverterService,
    {
      provide: ErrorHandler,
      useClass: ErrorHandlerService
    }
  ],
})
export class CoreModule {
  constructor(@Optional() @SkipSelf() parentModule: CoreModule) {
    throwIfAlreadyLoaded(parentModule, 'CoreModule');
  }
}
