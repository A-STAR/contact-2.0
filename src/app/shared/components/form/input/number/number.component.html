<label>
  <div *ngIf="label" class="form-label">
    {{ label | translate }}
    <b *ngIf="required" class="text-danger">*</b>
  </div>
  <div>
    <input
      class="form-control"
      type="text"
      [disabled]="disabled"
      [ngModel]="value"
      (focus)="onFocus()"
      (focusout)="onFocusOut()"
      (keydown)="onKeyDown($event)"
      (ngModelChange)="onChange($event)"
    />
    <i class="fa fa-angle-up" (click)="onIncrementClick()"></i>
    <i class="fa fa-angle-down" (click)="onDecrementClick()"></i>
  </div>
  <div>
    <span *ngIf="errors">{{ errors | json }}</span>
  </div>
</label>
