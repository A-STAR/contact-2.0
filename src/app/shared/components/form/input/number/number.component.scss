@import '../../../../styles/app/variables.scss';

:host {
  display: block;
}

.input-container {
  position: relative;
}

i {
  color: $gray;
  cursor: pointer;
  line-height: 0.5;
  position: absolute;
  right: 5px;
  transition: color 0.2s ease;

  &:before {
    display: inline-block;
  }

  &.fa-angle-up {
    top: 5px;
  }

  &.fa-angle-down {
    bottom: 5px;
  }
}

.input-container:hover > input:not(:disabled) {
  & ~ i {
    color: $text-light;
  }
  & ~ i:hover {
    color: $link-color;
  }
}
