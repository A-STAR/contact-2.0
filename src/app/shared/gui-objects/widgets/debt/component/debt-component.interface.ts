export interface IDebtComponent {
  id: number;
  typeCode: number;
  sum: number;
  currencyId: number;
}
