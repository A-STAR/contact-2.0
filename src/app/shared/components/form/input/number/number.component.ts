import {
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  Component,
  ElementRef,
  Input,
  Renderer2,
  forwardRef,
} from '@angular/core';
import { ControlValueAccessor, NG_VALIDATORS, NG_VALUE_ACCESSOR, Validator } from '@angular/forms';
import { range } from '@app/core/utils';
import { defaultTo } from 'ramda';
import { isNumber } from 'util';

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  providers: [
    {
      provide: NG_VALUE_ACCESSOR,
      useExisting: forwardRef(() => NumberComponent),
      multi: true,
    },
    {
      provide: NG_VALIDATORS,
      useExisting: forwardRef(() => NumberComponent),
      multi: true,
    },
  ],
  selector: 'app-number',
  styleUrls: [ './number.component.scss' ],
  templateUrl: './number.component.html'
})
export class NumberComponent implements ControlValueAccessor, Validator {
  private static readonly ALLOWED_KEYS = {
    main: [
      ...range(0, 9).map(String),
      '.',
      '-',
    ],
    aux: [
      'Backspace',
      'Delete',
      'ArrowLeft',
      'ArrowRight',
      'Home',
      'End',
    ],
  };

  private static readonly MAX_LENGTH = 12;
  @Input() errors: any;
  @Input() label: string;
  @Input() min: number;
  @Input() max: number;
  @Input() positive = false;
  @Input() required = false;
  @Input() precision = 0;

  @Input()
  set isReadonly(value: boolean) {
    this.readonly = this.setDefault(value, this.readonly);
    this.required = this.readonly ? false : this.required;
  }

  @Input()
  set isDisabled(value: boolean) {
    this.disabled = this.setDefault(value, this.disabled);
    this.required = this.disabled ? false : this.required;
  }

  @Input() set step(step: number | string) {
    this._step = Number(step);
  }

  disabled = false;
  readonly = false;
  value: number;

  private _step = 1;

  private wheelListenter: () => void;

  constructor(
    private cdRef: ChangeDetectorRef,
    private elRef: ElementRef,
    private renderer: Renderer2,
  ) {}

  writeValue(value: number): void {
    this.value = this.precision && isNumber(value) ? Number(value.toFixed(this.precision)) : value;
    this.cdRef.markForCheck();
  }

  registerOnChange(fn: Function): void {
    this.propagateChange = fn;
  }

  registerOnTouched(fn: Function): void {
    this.propagateTouch = fn;
  }

  setDisabledState(disabled: boolean): void {
    this.disabled = disabled;
  }

  validate(): any {
    switch (true) {
      case this.required && this.value == null:
        return { required: true };
      case this.positive && isNumber(this.value) && this.value <= 0:
        return { positive: true };
      case !this.isMinValid(this.value):
        return { min: { minValue: this.min } };
      case !this.isMaxValid(this.value):
        return { max: { maxValue: this.max } };
      default:
        return null;
    }
  }

  onKeyDown(event: KeyboardEvent): void {
    const { ctrlKey, key, target } = event;
    const { value } = target as HTMLInputElement;
    const { main, aux } = NumberComponent.ALLOWED_KEYS;
    if ((!main.includes(key) || value.length >= NumberComponent.MAX_LENGTH) && !aux.includes(key) && !ctrlKey) {
      event.preventDefault();
    }
  }

  onChange(value: string): void {
    const valueAsNumber = value === '' ? null : Number(value);
    const newValue = isNaN(valueAsNumber) ? null : valueAsNumber;
    this.update(newValue);
  }

  onFocus(): void {
    this.wheelListenter = this.renderer.listen(this.elRef.nativeElement, 'wheel', event => this.onWheel(event));
  }

  onFocusOut(): void {
    this.wheelListenter();
    this.propagateTouch();
  }

  onWheel(event: WheelEvent): void {
    if (!(this.disabled || this.readonly)) {
      event.preventDefault();
      const value = (this.value || 0) - this._step * Math.sign(event.deltaY);
      if (this.isMinValid(value) && this.isMaxValid(value)) {
        this.update(value);
      }
    }
  }

  onIncrementClick(): void {
    if (!(this.disabled || this.readonly)) {
      const value = (this.value || 0) + this._step;
      if (this.isMinValid(value) && this.isMaxValid(value)) {
        this.update(value);
      }
    }
  }

  onDecrementClick(): void {
    if (!(this.disabled || this.readonly)) {
      const value = (this.value || 0) - this._step;
      if (this.isMinValid(value) && this.isMaxValid(value)) {
        this.update(value);
      }
    }
  }

  private setDefault(value: boolean, defaultValue: boolean): boolean {
    return defaultTo(defaultValue)(value);
  }

  private update(value: number): void {
    this.value = this.precision ? Number(value.toFixed(this.precision)) : value;
    this.propagateChange(value);
    this.cdRef.markForCheck();
  }

  private isMinValid(value: number): boolean {
    return this.value == null || this.min == null || value >= this.min;
  }

  private isMaxValid(value: number): boolean {
    return this.value == null || this.max == null || value <= this.max;
  }

  private propagateChange: Function = () => {};
  private propagateTouch: Function = () => {};
}
