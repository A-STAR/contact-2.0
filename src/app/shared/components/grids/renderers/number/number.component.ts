import { ChangeDetectionStrategy, Component } from '@angular/core';
import { ICellRendererParams } from 'ag-grid/main';
import { ICellRendererAngularComp } from 'ag-grid-angular';

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'app-grid-number-renderer',
  template: `{{ value }}`,
})
export class NumberRendererComponent implements ICellRendererAngularComp {
  private params: ICellRendererParams;

  agInit(params: ICellRendererParams): void {
    this.params = params;
  }

  refresh(): boolean {
    return false;
  }

  get value(): string {
    return this.params.value;
  }
}
