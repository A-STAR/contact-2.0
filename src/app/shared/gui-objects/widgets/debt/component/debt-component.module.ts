import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

import { DebtComponentGridModule } from './grid/debt-component-grid.module';

import { DebtComponentService } from './debt-component.service';

@NgModule({
  imports: [
    DebtComponentGridModule,
    CommonModule,
  ],
  exports: [
    DebtComponentGridModule,
  ],
  providers: [
    DebtComponentService,
  ]
})
export class DebtComponentModule { }
