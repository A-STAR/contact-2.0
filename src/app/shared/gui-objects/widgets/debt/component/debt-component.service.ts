import { Injectable } from '@angular/core';
import { Observable } from 'rxjs/Observable';

import { IDebtComponent } from './debt-component.interface';

import { DataService } from '../../../../../core/data/data.service';
import { NotificationsService } from '../../../../../core/notifications/notifications.service';

@Injectable()
export class DebtComponentService {
  static MESSAGE_DEBT_COMPONENT_SAVED = 'MESSAGE_DEBT_COMPONENT_SAVED';

  constructor(
    private dataService: DataService,
    private notificationsService: NotificationsService,
  ) {}

  fetchAll(debtId: number): Observable<Array<IDebtComponent>> {
    return this.dataService
      .read('/debts/{debtId}/components', { debtId })
      .map(response => response.components)
      .catch(this.notificationsService.error('errors.default.read').entity('entities.debtComponents.gen.plural').dispatchCallback());
  }

  fetch(debtId: number, debtComponentId: number): Observable<IDebtComponent> {
    return this.dataService
      .read('/debts/{debtId}/components/{debtComponentId}', { debtId, debtComponentId })
      .map(response => response.components[0])
      .catch(this.notificationsService.error('errors.default.read').entity('entities.debtComponents.gen.singular').dispatchCallback());
  }

  create(debtId: number, debtComponent: IDebtComponent): Observable<void> {
    return this.dataService
      .create('/debts/{debtId}/components', { debtId }, debtComponent)
      .catch(this.notificationsService.error('errors.default.create').entity('entities.debtComponents.gen.singular').dispatchCallback());
  }

  update(debtId: number, debtComponentId: number, debtComponent: Partial<IDebtComponent>): Observable<void> {
    return this.dataService
      .update('/debts/{debtId}/components/{debtComponentId}', { debtId, debtComponentId }, debtComponent)
      .catch(this.notificationsService.error('errors.default.update').entity('entities.debtComponents.gen.singular').dispatchCallback());
  }

  delete(debtId: number, debtComponentId: number): Observable<void> {
    return this.dataService
      .delete('/debts/{debtId}/components/{debtComponentId}', { debtId, debtComponentId })
      .catch(this.notificationsService.error('errors.default.delete').entity('entities.debtComponents.gen.singular').dispatchCallback());
  }
}
