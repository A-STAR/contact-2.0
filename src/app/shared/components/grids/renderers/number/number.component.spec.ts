import { async as Async, ComponentFixture, TestBed } from '@angular/core/testing';
import { FormsModule } from '@angular/forms';
import { TranslateLoader, TranslateModule } from '@ngx-translate/core';
import { Observable } from 'rxjs/Observable';
import { of } from 'rxjs/observable/of';

import { NumberRendererComponent } from './number.component';

class TranslateLoaderMock {
  getTranslation(language: string): Observable<any> {
    return of([]);
  }
}

describe('NumberRendererComponent', () => {
  let fixture: ComponentFixture<NumberRendererComponent>;

  beforeEach(Async(() => {
    TestBed
      .configureTestingModule({
        imports: [
          FormsModule,
          TranslateModule.forRoot({
            loader: {
              provide: TranslateLoader,
              useClass: TranslateLoaderMock,
            },
          }),
        ],
        declarations: [
          NumberRendererComponent,
        ],
      })
      .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(NumberRendererComponent);
  });

  it('should render number renderer', () => {
    fixture.componentInstance.agInit({ value: 1 } as any);
    fixture.detectChanges();
    expect(fixture.nativeElement).toMatchSnapshot();
  });
});
