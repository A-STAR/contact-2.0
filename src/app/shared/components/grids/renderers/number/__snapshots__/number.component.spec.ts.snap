// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberRendererComponent should render number renderer 1`] = `
<div
  id="root0"
  ng-version="5.2.4"
>
  1
</div>
`;
